platform/src/main/java/org/hillview/main/BatchLogAnalysis.java

```diff
@@ -0,0 +1,188 @@
+import org.hillview.table.api.ITable;
+import org.hillview.utils.Converters;
+import org.hillview.utils.JsonList;
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.cli.*;
+
+
+public class BatchLogAnalysis {
+
+    private static class HeatmapData {
+        long[][] matrix;
+        JsonList<String> errorCodeLabels = new JsonList<>();
+        ArrayList<String> timeLabels = new ArrayList<>();
+    }
+
+    private static HeatmapData heatmapErrTime(FileSetDescription desc, LocalDataSet<Empty> local, int numOfTimestampBuckets) {
+        IMap<Empty, List<IFileReference>> finder = new FindFilesMapper(desc);
+        IDataSet<IFileReference> found = local.blockingFlatMap(finder);
+        IMap<IFileReference, ITable> loader = new LoadFilesMapper();
+        IDataSet<ITable> table = found.blockingMap(loader);
+        ExtractValueFromKeyMap evkm = new ExtractValueFromKeyMap("errorCode", "StructuredData", "errorCode", -1);
+        IDataSet<ITable> table1 = table.blockingMap(evkm);
+
+        DoubleDataRangeSketch rangeSketch = new DoubleDataRangeSketch("Timestamp");
+        DataRange dataRange = table1.blockingSketch(rangeSketch);
+        DoubleHistogramBuckets bucketsTimestamp = new DoubleHistogramBuckets(dataRange.min, dataRange.max, numOfTimestampBuckets);
+
+        SampleDistinctElementsSketch sampleSketch = new SampleDistinctElementsSketch("errorCode", 0, 200);
+        MinKSet<String> samples = table1.blockingSketch(sampleSketch);
+        JsonList<String> leftBoundaries = samples.getLeftBoundaries(200);
+        StringHistogramBuckets bucketsErrorCode = new StringHistogramBuckets(leftBoundaries.toArray(new String[0]));
+
+        HeatmapSketch heatmapSketch = new HeatmapSketch(bucketsTimestamp, bucketsErrorCode, "Timestamp", "errorCode", 1.0, 0);
+        Heatmap heatmap = table1.blockingSketch(heatmapSketch);
+        HeatmapData heatmapData = new HeatmapData();
+        int numOfBucketsD1 = heatmap.getNumOfBucketsD1();
+        int numOfBucketsD2 = heatmap.getNumOfBucketsD2();
+        heatmapData.matrix = new long[numOfBucketsD1][numOfBucketsD2];
+        for (int i = 0; i<numOfBucketsD1; i++){
+            for (int j = 0; j<numOfBucketsD2; j++){
+                heatmapData.matrix[i][j] = heatmap.getCount(i, j);
+            }
+        }
+        heatmapData.errorCodeLabels = leftBoundaries;
+        for (int x = 0; x < numOfTimestampBuckets; x++){
+            double time = dataRange.min + x * (dataRange.max - dataRange.min) / numOfTimestampBuckets;
+            Instant instantTime = Converters.toDate(time);
+            String stringDate = Converters.toString(instantTime);
+            heatmapData.timeLabels.add(stringDate);
+        }
+        return heatmapData;
+    }
+
+    private static void saveHeatmapToFile(HeatmapData heatmapData, String filepath) {
```

Comment:
For this function you should document the file format you are using. If this is a data frame, just say that.